// @keystonejs/field-content/dist/chunk-4abd0091.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

require("@babel/runtime/helpers/objectWithoutProperties");

require("@babel/runtime/helpers/extends");

require('@emotion/core');

var __chunk_2 = require('./chunk-246895b9.cjs.js');

var icons = require('@arch-ui/icons');

require('@arch-ui/theme');

require('@arch-ui/color-utils');

require('@arch-ui/tooltip');

require('@arch-ui/typography');

var __chunk_3 = require('./chunk-c7a01e38.cjs.js');

var React = require('react');

var React__default = _interopDefault(React);

var paragraph_js = require('./chunk-b5f8afc7.cjs.js');

var listItem_js = require('./chunk-59088f9e.cjs.js');

let handleListButtonClick = (editor, editorState, type) => {
  let isListItem = __chunk_2.hasBlock(editorState, listItem_js.type);

  let isOrderedList = __chunk_2.hasAncestorBlock(editorState, type);

  let otherListType = type === 'ordered-list' ? 'unordered-list' : 'ordered-list';

  if (isListItem && isOrderedList) {
    editor.setBlocks(paragraph_js.type);
    editor.unwrapBlock(type);
  } else if (isListItem) {
    editor.unwrapBlock(otherListType);
    editor.wrapBlock(type);
  } else {
    editor.setBlocks(listItem_js.type).wrapBlock(type);
  }

  editor.focus();
};

let type = 'ordered-list';

function ToolbarElement(_ref) {
  let {
    editor,
    editorState
  } = _ref;
  return React.createElement(__chunk_3.ToolbarButton, {
    label: "Ordered List",
    icon: React.createElement(icons.ListOrderedIcon, null),
    isActive: __chunk_2.hasAncestorBlock(editorState, type),
    onClick: () => {
      handleListButtonClick(editor, editorState, type);
    }
  });
}

function Node(_ref2) {
  let {
    attributes,
    children
  } = _ref2;
  return React.createElement("ol", attributes, children);
}

let getPlugins = () => [{
  onKeyDown(event, editor, next) {
    // make it so when you press enter in an empty list item,
    // the block type will change to a paragraph
    if (event.keyCode === 13 && __chunk_2.hasAncestorBlock(editor.value, type) && editor.value.focusText.text === '') {
      editor.setBlocks(paragraph_js.type).unwrapBlock(type);
      return;
    }

    next();
  }

}];

const getSchema = () => ({
  nodes: [{
    match: {
      type: listItem_js.type
    },
    min: 0
  }],

  normalize(editor, error) {
    switch (error.code) {
      case 'child_type_invalid':
        {
          editor.unwrapBlockByKey(error.node.key, type);
          return;
        }
    }
  }

});

exports.type = type;
exports.ToolbarElement = ToolbarElement;
exports.Node = Node;
exports.getPlugins = getPlugins;
exports.getSchema = getSchema;
